doc/docs/master/how-tos/enhanced-pipeline-workflow.md

```diff
@@ -0,0 +1,121 @@
+- language: A string specifying what language builder to use (see below.) Only works with official builders.
+- image: A string specifying what builder image to use, if a non-official builder is desired.
+
+Below is a Python example of a build pipline.
+
+```json
+{
+  "pipeline": {
+    "name": "map"
+  },
+  "description": "A pipeline that tokenizes scraped pages and appends counts of words to corresponding files.",
+  "transform": {
+    "build": {
+      "language": "python",
+      "path": "./source"
+    }
+  },
+  "input": {
+    "pfs": {
+      "repo": "scraper",
+      "glob": "/*"
+    }
+  }
+}
+```
+
+A build pipeline can be submitted the same way as any other pipeline, for example:
+
+```bash
+pachctl update pipeline -f <pipeline name> 
+```
+
+When submitted, the following actions occur: 
+
+1. All files (code, etc.) are copied from the build path to a PFS repository, `<pipeline name>_build`, which we can think of as the source code repository. In the case above, everything in `./source` would be copied to to the PFS `<pipeline name>_build` repository.
+
+1. Starts a pipeline called `<pipeline name>_build` that reads the files uploaded and runs the `build.sh` script to pull dependencies and compile any requirements from the source code. 
+
+1. Creates (or updates if it already exists) the running pipeline `<pipeline name>` with the the newly creates source files and built assets. 
+
+!!! note
+      You can optionally specify a `.pachignore` file to prevent certain files from getting pushed to this repo. 
+
+The updated pipeline contains the following PFS repos mapped in as inputs:
+
+1. `/pfs/source` - source code that is required for running the pipeline. There are language-specific requirements on the structure of this source code mentioned in the [Python Builder](#python-builder) section.
+
+1. `/pfs/build` - any artifacts resulting from the build process. 
+
+1. `/pfs/<input(s)>` - any inputs specified in the pipeline specification
+
+### Builders
+Either `language` or `image` must be specified in the pipeline spec. The langauge parameter informs the pipeline of what base image to use and already contain implementations for `build.sh` and `run.sh`. Go and Python "builders" have a standarized implementation, only requiring setting the `language` parameter in the pipeline specification.  
+
+If `language` is not specified, then the build pipeline uses the `transform.image` container (if `transform.image` is not specified the builder image is used). Similarly, if `transform.cmd` is not specified, it will default to running `sh /pfs/build/run.sh`, which should be provided via the build pipeline.
+
+#### Python Builder
+
+The Python builder requires a file structure similar to the following:
+
+```tree
+./map
+├── source
+│   ├── requirements.txt
+│   ├── main.py
+│   ├── build.sh (optional)
+│   └── run.sh (optional)
```

Comment:
User-specified `build.sh` and `run.sh` are actually not respected (this is confusingly different from python_pachyderm's `create_python_pipeline`, which does respect them.)

Reply:
Ah, interesting. Is this correct: 
- If you are using the python builder, then everything is ignored except the `requirements.txt` and `main.py`.
- If you change the builder image, then you must provide a `build.sh` and a `run.sh`.

And I didn't include the `create_python_pipeline` method here since it is python-pachyderm specific. Should we add that to the docs over there or do we think that it will change over time?

Reply:
I updated this in the most recent commit. I'll wait to see if that resolves the issue before resolving this comment.

Reply:
Everything isn't ignore per se, because you can include, e.g. a directory that you import from `main.py`.

The new docs aren't quite right because on the build step-enabled pipelines, having a `build.sh` or `run.sh` in the source directory _has no effect_, unless the language builder uses them (which the existing ones do not.)